etk-asm/src/asm.rs

```diff
@@ -353,58 +353,43 @@ impl Assembler {
     }
 
     /// Feed a single instruction into the `Assembler`.
-    ///
-    /// Returns the number of bytes that can be collected with [`Assembler::take`]
-    pub fn push<O>(&mut self, rop: O) -> Result<usize, Error>
+    fn push<O>(&mut self, rop: &O) -> Result<usize, Error>
     where
         O: Into<RawOp>,
     {
         let rop = rop.into();
 
-        self.declare_content(&rop)?;
-
-        // Expand instruction macros immediately. We do this here because it's the same process
-        // regardless if we `push_read` or `push_pending` -- in fact, `expand_macro` pushes each op
-        // individually which calls the correct unchecked push.
-        if let RawOp::Op(AbstractOp::Macro(ref m)) = rop {
-            self.expand_macro(&m.name, &m.parameters)?;
-            return Ok(self.ready.len());
-        }
-
-        self.push_unchecked(rop)?;
-        Ok(self.ready.len())
-    }
-
-    fn push_unchecked(&mut self, rop: RawOp) -> Result<(), Error> {
-        if self.pending.is_empty() && self.pending_len.is_some() {
-            self.push_ready(rop)
-        } else {
-            self.push_pending(rop)
-        }
-    }
+        self.declare_label(&rop)?;
 
-    fn push_ready(&mut self, rop: RawOp) -> Result<(), Error> {
         match rop {
             RawOp::Op(AbstractOp::Label(label)) => {
+                let mut dst = 0;
+                for ul in self.undeclared_labels.iter() {
+                    if (ul.label == label) & ul.dynamic_push {
+                        // Compensation in case label was dynamically pushed.
+                        let mut tmp =
+                            ((self.concrete_len as f32 - ul.position as f32) / 256.0).floor();
+
+                        // Size already accounted for %push(label) was 2.
+                        if tmp > 1.0 {
+                            tmp -= 1.0;
+                        }
+                        dst = cmp::max(tmp as usize, dst);
```

Comment:
Something smells off about this, but I can't quite put my finger on it. Let me try a few things, and I'll be back.

Reply:
Sure! Let me know if you have any questions or if I can help you in any way. Basically that part of the code takes care of compensating the value of the labels. If a dynamic push was used to define the label in question, in that instance its position is already known, so it "stretches" the value to reach the final position. As all intermediate macros were already expanded because they are known beforehand, there should be no problem in setting their value correctly.

But if you find any interesting edge case, feel free to let me know and I can add the test to be sure it works.

Reply:
Oh, I see, I didn't realise that labels could be defined that way. Excellent catch. I'll fix the code as quickly as possible.

Reply:
My thought is that these two programs must assemble to the same thing:

<table>
<tr>
<th>
</th>
<th>Constant Push</th>
<th>Variable Push</th>
</tr>
<tr>
<th>Input</th>
<td>

```
push2 label + 254                   
label:
pc
```

</td>
<td>

```
%push(label + 254)                   
label:
pc
```

</td>
</tr>

<tr>
<th>Output</th>
<td>

```
61010158
```

</td>
<td>

```
61010058
```

</td>
</tr>
</table>

I assumed the constant push was correct, but regardless, they should be the same.

Reply:
Oh, I see. I'll check it out. Thanks!

Reply:
Just to be sure. This should be a valid program too, right?

```
%push(label1 + label2 + 254)
label:
pc
label2:
pc
```

Reply:
Nothing out of the ordinary; I was just working on the previous case and noticed that this one wasn't included in my solution, and I thought it should be. I wanted to double-check. I'll figure out a solution that covers both, but it might mean changing some parts you've already looked at. Sorry about that.

Reply:
Solved. There were many intermediate steps, so I simplified the code a bit and renamed things for better understanding.